### DISCUSSION TITLE
CreateModuleCatalog() in PrismBootstrapper - Error: Unable to cast transparent proxy to type 'InnerModuleInfoLoader'

### DISCUSSION THREAD

**Discussion by melsawy93** [PrismLibrary/Prism#2580] [WPF] [answered]
I am creating a plugin that creates a button in a software I am using. I debugged the code I have and my code crashes (at the bootstrapper.Run()) when running showing the following message

Exception thrown: 'System.InvalidCastException' in Prism.Wpf.dll An exception of type 'System.InvalidCastException' occurred in Prism.Wpf.dll but was not handled in user code Unable to cast transparent proxy to type 'InnerModuleInfoLoader'.

The entry point of my plugin contains:

    Bootstrapper bootstrapper = new Bootstrapper();
    bootstrapper.Run();

My bootstrapper.cs class:
 ```
public class Bootstrapper : PrismBootstrapperBase
    {
        protected override IContainerExtension CreateContainerExtension()
        {
            return new UnityContainerExtension();
        }

        protected override DependencyObject CreateShell()
        {
            return Container.Resolve<MainWindow>();
        }

        protected override void RegisterTypes(IContainerRegistry containerRegistry)
        {
            containerRegistry.Register<MainWindowViewModel>();
        }

        protected override void ConfigureViewModelLocator()
        {
            base.ConfigureViewModelLocator();
        }

        protected override IModuleCatalog CreateModuleCatalog()
        {
            // Use local directory as module catalog
            return new DirectoryModuleCatalog { ModulePath = "Modules" };
        }

    }
```
My modules have a post-build command to copy the .dll in the bin/debug/Modules folder for my plugin and I have checked the folder myself after building (its there). I really dont know whats wrong.

My module class:
```

using Combinito.Modules.LCGenerator.Views;
using Prism.Ioc;
using Prism.Modularity;
using Prism.Regions;

namespace Combinito.Modules.LCGenerator
{
    /// <summary>
    /// Register components of module with Unity/Prism
    /// </summary>
    [Module(ModuleName = "ModuleA")]
    public class LCGeneratorModule : IModule
    {
        
        public void Initialize()
        {

        }
        public void OnInitialized(IContainerProvider containerProvider)
        {
 
        }

        public void RegisterTypes(IContainerRegistry containerRegistry)
        {
            
        }
    }
}
```
This is similar to : https://stackoverflow.com/questions/8158512/error-running-a-wpf-prism-app-from-an-arcgis-desktop-add-in-button-unable-to-lo

However I dont really understand the solution they mention there. Prism.dll is already in the plugin's bin/debug folder..

**Comment by melsawy93** [comment]
I have been smashing my head at this for a while. I downloaded Fusion++ to take a look at the .dll being loaded. I noticed just when the error happens I get an error loading Prism.wpf.dll. Even though its in the bin/debug folder but for reason the compiler or whatever is looking into the folder of the host software (hosting the plugin). Please see below:

```
*** Assembly Binder Log Entry  (10/13/2021 @ 10:30:40 PM) ***

The operation failed.
Bind result: hr = 0x80070002. The system cannot find the file specified.

Assembly manager loaded from:  C:\Windows\Microsoft.NET\Framework64\v4.0.30319\clr.dll
Running under executable  C:\Program Files\Computers and Structures\ETABS 19\ETABS.exe
--- A detailed error log follows. 

=== Pre-bind state information ===
LOG: DisplayName = Prism.Wpf, Version=8.1.97.5141, Culture=neutral, PublicKeyToken=40ee6c3a2184dc59
 (Fully-specified)
LOG: Appbase = file:///C:/Program Files/Computers and Structures/ETABS 19/
LOG: Initial PrivatePath = NULL
LOG: Dynamic Base = NULL
LOG: Cache Base = NULL
LOG: AppName = ETABS.exe
Calling assembly : (Unknown).
===
LOG: This bind starts in default load context.
LOG: Using application configuration file: C:\Program Files\Computers and Structures\ETABS 19\ETABS.exe.Config
LOG: Using host configuration file: 
LOG: Using machine configuration file from C:\Windows\Microsoft.NET\Framework64\v4.0.30319\config\machine.config.
LOG: GAC Lookup was unsuccessful.
LOG: Attempting download of new URL file:///C:/Program Files/Computers and Structures/ETABS 19/Prism.Wpf.DLL.
LOG: Attempting download of new URL file:///C:/Program Files/Computers and Structures/ETABS 19/Prism.Wpf/Prism.Wpf.DLL.
LOG: Attempting download of new URL file:///C:/Program Files/Computers and Structures/ETABS 19/Prism.Wpf.EXE.
LOG: Attempting download of new URL file:///C:/Program Files/Computers and Structures/ETABS 19/Prism.Wpf/Prism.Wpf.EXE.
LOG: All probing URLs attempted and failed.
```

**Comment by brianlagunas** [comment]
Hard to say as your debugging environment is dependent on Fusion++. All I can say is that you need to know where Fusion is looking for plugin dependencies , and then make sure **ALL** your plugin dependencies are there. Even framework dependencies if not being pulled from the GAC.

**Reply by melsawy93** [reply]
Thanks alot for getting back to me. I have posted a more detailed question here: https://stackoverflow.com/questions/69564696/unable-to-cast-transparent-proxy-to-type-assembly-loading-issue?noredirect=1#comment122969871_69564696. Can you please take a quick look if you got a second and give me any hints? Im new to this and any guidance would be appreciated. What is puzzling me is that without creating any modules with Prism, a simple Prism based plugin works. As soon as I add a module it just doesnt.

**Reply by melsawy93** [reply]
I have been reading about how dll are loaded; You will notice fusion only does 4 attempts to get Prism.Wpf; why doesnt it do the 5th attempt which should be in my plugin folder? I am comparing it to the first log shown by Fusion++:

```
`LOG: Appbase = file:///C:/Program Files/Computers and Structures/ETABS 19/
LOG: Initial PrivatePath = NULL
LOG: Dynamic Base = NULL
LOG: Cache Base = NULL
LOG: AppName = ETABS.exe
Calling assembly : Combinito, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null.
===
LOG: This bind starts in LoadFrom load context.
WRN: Native image will not be probed in LoadFrom context. Native image will only be probed in default load context, like with Assembly.Load().
LOG: Using application configuration file: C:\Program Files\Computers and Structures\ETABS 19\ETABS.exe.Config
LOG: Using host configuration file: 
LOG: Using machine configuration file from C:\Windows\Microsoft.NET\Framework64\v4.0.30319\config\machine.config.
LOG: Post-policy reference: Prism.Wpf, Version=8.1.97.5141, Culture=neutral, PublicKeyToken=40ee6c3a2184dc59
LOG: GAC Lookup was unsuccessful.
LOG: Attempting download of new URL file:///C:/Program Files/Computers and Structures/ETABS 19/Prism.Wpf.DLL.
LOG: Attempting download of new URL file:///C:/Program Files/Computers and Structures/ETABS 19/Prism.Wpf/Prism.Wpf.DLL.
LOG: Attempting download of new URL file:///C:/Program Files/Computers and Structures/ETABS 19/Prism.Wpf.EXE.
LOG: Attempting download of new URL file:///C:/Program Files/Computers and Structures/ETABS 19/Prism.Wpf/Prism.Wpf.EXE.
LOG: Attempting download of new URL file:///C:/Users/moustafa.el-sawy/source/repos/Combinito/Combinito/bin/Debug/Prism.Wpf.DLL.
LOG: Assembly download was successful. Attempting setup of file: C:\Users\moustafa.el-sawy\source\repos\Combinito\Combinito\bin\Debug\Prism.Wpf.dll`
```

Also why is it loaded twice? The person commenting on my post might provide valuable information (that I dont understand). Again, I appreciate your help. I just need guidance and I am willing to put the work to fix this. If there is a better way to debug this please let me know and I am willing to do it and show you the output. 

Edit: I created my plugin following your create outlook series on youtube with minor modifications. Would you like me to upload the code? its very small sample plugin with a single view.

**Reply by brianlagunas** [reply]
If it's just the module failing, then don't use the DirectoryModule and instead just add it via a hard reference in code. If it works, then you know you have an issue with either your paths or your dependencies.

We offer paid support if you really need it.

**Reply by melsawy93** [reply]
Okay I will look into how to add a hard reference and see if that works (if theres an example you can share I would appreciated).

I have seen that you offer paid support but I was always under the impression that it costs alot of money. I dont mind paying for support if I can afford it given that the plugin I am developing is for a hobby and not commercially sold (more for personal development). Is there a rate or place I can see the price for support?

Edit: just saw I can put in contact us request in AvantiPoint. To avoid wasting your time; is there a rate for non-enterprise single users who would just need help (hourly rate maybe as I have issues?). If there is I would be glad to put in a request to discuss further.

Thanks for your time!

**Reply by brianlagunas** [reply]
Samples can be found here: https://github.com/PrismLibrary/Prism-Samples-Wpf

I don't know AvantiPoint's rates. I would just reach out and ask them.

**Reply by melsawy93** [reply]
I will reach out, thank you!

I have seen these samples already and tried it. When I try to implement the following:

```
protected override void ConfigureModuleCatalog(IModuleCatalog moduleCatalog)

        {
            moduleCatalog.AddModule<LCombGeneratorModule>();
        }

```

I am getting the following error and I am not able to build a solution:

> Severity	Code	Description	Project	File	Line	Suppression State
Error	CS0121	The call is ambiguous between the following methods or properties: 'Prism.Modularity.IModuleCatalogExtensions.AddModule<T>(Prism.Modularity.IModuleCatalog, Prism.Modularity.InitializationMode)' and 'Prism.Modularity.IModuleCatalogExtensions.AddModule<T>(Prism.Modularity.IModuleCatalog, Prism.Modularity.InitializationMode)'	Combinito	C:\Users\moustafa.el-sawy\source\repos\Combinito\Combinito\Bootstrapper.cs	44	Active

Thanks again.

**Reply by brianlagunas** [reply]
Then you might have other issues in your code. You don't have to use the generic. You can just use the method `moduleCatalog.AddModule(...)`

**Reply by melsawy93** [reply]
Just made the change and I am getting another error.
```
`protected override void ConfigureModuleCatalog(IModuleCatalog moduleCatalog)
        {
            moduleCatalog.AddModule(LCombGeneratorModule);
        }
`
```

> Severity	Code	Description	Project	File	Line	Suppression State
Error	CS0119	'LCombGeneratorModule' is a type, which is not valid in the given context	Combinito	C:\Users\moustafa.el-sawy\source\repos\Combinito\Combinito\Bootstrapper.cs	44	Active

**Reply by brianlagunas** [reply]
you need to provide the type, not the class

**Reply by melsawy93** [reply]
omg, thank you so much. It works!! I changed it to typeof(...)!! I have learnt alot debugging this!

**Comment by melsawy93** [comment, marked as answer]
Anyone facing same issue: I had to used ConfigureModuleCatalog() instead of CreateModuleCatalog(). I have yet to know why CreateModuleCatalog() wasnt working for me.

Final code:
```
`        protected override void ConfigureModuleCatalog(IModuleCatalog moduleCatalog)

        {
            moduleCatalog.AddModule(typeof(LCombGeneratorModule));
        }
`

```

Edit: thanks to Brian's help.
